File: dagayn/parser/languages/r.py
"""R construct extraction for the CodeParser walker."""

from __future__ import annotations

from typing import Optional

from .._base.protocol import CodeParser
from .._base.test_detection import is_test_function as _is_test_function
from .._base.types import EdgeInfo, NodeInfo


def handle_node(
    parser: "CodeParser",
    child,
    node_type: str,
    source: bytes,
    language: str,
    file_path: str,
    nodes: list[NodeInfo],
    edges: list[EdgeInfo],
    enclosing_class: Optional[str],
    enclosing_func: Optional[str],
    import_map: Optional[dict[str, str]],
    defined_names: Optional[set[str]],
    _depth: int,
) -> bool:
    return _extract_r_constructs(
        parser,
        child,
        node_type,
        source,
        language,
        file_path,
        nodes,
        edges,
        enclosing_class,
        enclosing_func,
        import_map,
        defined_names,
    )


def _extract_r_constructs(
    parser: "CodeParser",
    child,
    node_type: str,
    source: bytes,
    language: str,
    file_path: str,
    nodes: list[NodeInfo],
    edges: list[EdgeInfo],
    enclosing_class: Optional[str],
    enclosing_func: Optional[str],
    import_map: Optional[dict[str, str]],
    defined_names: Optional[set[str]],
) -> bool:
    """Handle R-specific AST nodes (assignments and class-defining calls).

    Returns True if the child was fully handled and should be skipped
    by the main loop.
    """
    # R: function definitions via assignment
    if node_type == "binary_operator":
        handled = _handle_r_binary_operator(
            parser,
            child,
            source,
            language,
            file_path,
            nodes,
            edges,
            enclosing_class,
            enclosing_func,
            import_map,
            defined_names,
        )
        if handled:
            return True

    # R: setClass/setRefClass/setGeneric calls and imports
    if node_type == "call":
        handled = _handle_r_call(
            parser,
            child,
            source,
            language,
            file_path,
            nodes,
            edges,
            enclosing_class,
            enclosing_func,
            import_map,
            defined_names,
        )
        if handled:
            return True

    return False


# ------------------------------------------------------------------
# R-specific helpers
# ------------------------------------------------------------------


def _r_call_func_name(call_node) -> Optional[str]:
    """Extract the function name from an R call node."""
    for child in call_node.children:
        if child.type in ("identifier", "namespace_operator"):
            return child.text.decode("utf-8", errors="replace")
    return None


def _r_first_string_arg(call_node) -> Optional[str]:
    """Extract the first string argument value from an R call node."""
    for child in call_node.children:
        if child.type == "arguments":
            for arg in child.children:
                if arg.type == "argument":
                    for sub in arg.children:
                        if sub.type == "string":
                            for sc in sub.children:
                                if sc.type == "string_content":
                                    return sc.text.decode("utf-8", errors="replace")
            break
    return None


def _r_iter_args(call_node):
    """Yield (name_str, value_node) pairs from an R call's arguments."""
    for child in call_node.children:
        if child.type != "arguments":
            continue
        for arg in child.children:
            if arg.type != "argument":
                continue
            has_eq = any(sub.type == "=" for sub in arg.children)
            if has_eq:
                name = None
                value = None
                for sub in arg.children:
                    if sub.type == "identifier" and name is None:
                        name = sub.text.decode("utf-8", errors="replace")
                    elif sub.type not in ("=", ","):
                        value = sub
                yield (name, value)
            else:
                for sub in arg.children:
                    if sub.type not in (",",):
                        yield (None, sub)
                        break
        break


def _r_find_named_arg(call_node, arg_name: str):
    """Find a named argument's value node in an R call."""
    for name, value in _r_iter_args(call_node):
        if name == arg_name:
            return value
    return None


# ------------------------------------------------------------------
# R-specific handlers
# ------------------------------------------------------------------


def _handle_r_binary_operator(
    parser: "CodeParser",
    node,
    source: bytes,
    language: str,
    file_path: str,
    nodes: list[NodeInfo],
    edges: list[EdgeInfo],
    enclosing_class: Optional[str],
    enclosing_func: Optional[str],
    import_map: Optional[dict[str, str]],
    defined_names: Optional[set[str]],
) -> bool:
    """Handle R binary_operator nodes: name <- function(...) { ... }."""
    children = node.children
    if len(children) < 3:
        return False

    left, op, right = children[0], children[1], children[2]
    if op.type not in ("<-", "="):
        return False

    if right.type == "function_definition" and left.type == "identifier":
        name = left.text.decode("utf-8", errors="replace")
        is_test = _is_test_function(name, file_path)
        kind = "Test" if is_test else "Function"
        qualified = parser._qualify(name, file_path, enclosing_class)
        params = parser._get_params(right, language, source)

        nodes.append(
            NodeInfo(
                kind=kind,
                name=name,
                file_path=file_path,
                line_start=right.start_point[0] + 1,
                line_end=right.end_point[0] + 1,
                language=language,
                parent_name=enclosing_class,
                params=params,
                is_test=is_test,
            )
        )

        container = (
            parser._qualify(enclosing_class, file_path, None) if enclosing_class else file_path
        )
        edges.append(
            EdgeInfo(
                kind="CONTAINS",
                source=container,
                target=qualified,
                file_path=file_path,
                line=right.start_point[0] + 1,
            )
        )

        parser._extract_from_tree(
            right,
            source,
            language,
            file_path,
            nodes,
            edges,
            enclosing_class=enclosing_class,
            enclosing_func=name,
            import_map=import_map,
            defined_names=defined_names,
        )
        return True

    if right.type == "call" and left.type == "identifier":
        call_func = _r_call_func_name(right)
        if call_func in ("setRefClass", "setClass", "setGeneric"):
            assign_name = left.text.decode("utf-8", errors="replace")
            return _handle_r_class_call(
                parser,
                right,
                source,
                language,
                file_path,
                nodes,
                edges,
                enclosing_class,
                enclosing_func,
                import_map,
                defined_names,
                assign_name=assign_name,
            )

    return False


def _handle_r_call(
    parser: "CodeParser",
    node,
    source: bytes,
    language: str,
    file_path: str,
    nodes: list[NodeInfo],
    edges: list[EdgeInfo],
    enclosing_class: Optional[str],
    enclosing_func: Optional[str],
    import_map: Optional[dict[str, str]],
    defined_names: Optional[set[str]],
) -> bool:
    """Handle R call nodes for imports and class definitions."""
    func_name = _r_call_func_name(node)
    if not func_name:
        return False

    if func_name in ("library", "require", "source"):
        imports = parser._extract_import(node, language, source)
        for imp_target in imports:
            edges.append(
                EdgeInfo(
                    kind="IMPORTS_FROM",
                    source=file_path,
                    target=imp_target,
                    file_path=file_path,
                    line=node.start_point[0] + 1,
                )
            )
        return True

    if func_name in ("setRefClass", "setClass", "setGeneric"):
        return _handle_r_class_call(
            parser,
            node,
            source,
            language,
            file_path,
            nodes,
            edges,
            enclosing_class,
            enclosing_func,
            import_map,
            defined_names,
        )

    # Module-scope R calls attribute to the File node.
    call_name = parser._get_call_name(node, language, source)
    if call_name:
        caller = (
            parser._qualify(enclosing_func, file_path, enclosing_class)
            if enclosing_func
            else file_path
        )
        target = parser._resolve_call_target(
            call_name,
            file_path,
            language,
            import_map or {},
            defined_names or set(),
        )
        edges.append(
            EdgeInfo(
                kind="CALLS",
                source=caller,
                target=target,
                file_path=file_path,
                line=node.start_point[0] + 1,
            )
        )
        edges.extend(parser._detect_cross_language_bridge(node, language, file_path, caller))

    parser._extract_from_tree(
        node,
        source,
        language,
        file_path,
        nodes,
        edges,
        enclosing_class=enclosing_class,
        enclosing_func=enclosing_func,
        import_map=import_map,
        defined_names=defined_names,
    )
    return True


def _handle_r_class_call(
    parser: "CodeParser",
    node,
    source: bytes,
    language: str,
    file_path: str,
    nodes: list[NodeInfo],
    edges: list[EdgeInfo],
    enclosing_class: Optional[str],
    enclosing_func: Optional[str],
    import_map: Optional[dict[str, str]],
    defined_names: Optional[set[str]],
    assign_name: Optional[str] = None,
) -> bool:
    """Handle setClass/setRefClass/setGeneric calls -> Class nodes."""
    class_name = _r_first_string_arg(node) or assign_name
    if not class_name:
        return False

    qualified = parser._qualify(class_name, file_path, enclosing_class)
    nodes.append(
        NodeInfo(
            kind="Class",
            name=class_name,
            file_path=file_path,
            line_start=node.start_point[0] + 1,
            line_end=node.end_point[0] + 1,
            language=language,
            parent_name=enclosing_class,
        )
    )
    edges.append(
        EdgeInfo(
            kind="CONTAINS",
            source=file_path,
            target=qualified,
            file_path=file_path,
            line=node.start_point[0] + 1,
        )
    )

    methods_list = _r_find_named_arg(node, "methods")
    if methods_list is not None:
        extract_r_methods(
            parser,
            methods_list,
            source,
            language,
            file_path,
            nodes,
            edges,
            class_name,
            import_map,
            defined_names,
        )

    return True


def extract_r_methods(
    parser: "CodeParser",
    list_call,
    source: bytes,
    language: str,
    file_path: str,
    nodes: list[NodeInfo],
    edges: list[EdgeInfo],
    class_name: str,
    import_map: Optional[dict[str, str]],
    defined_names: Optional[set[str]],
) -> None:
    """Extract methods from a setRefClass methods = list(...) call."""
    for method_name, func_def in _r_iter_args(list_call):
        if not method_name or func_def is None:
            continue
        if func_def.type != "function_definition":
            continue

        qualified = parser._qualify(method_name, file_path, class_name)
        params = parser._get_params(func_def, language, source)
        nodes.append(
            NodeInfo(
                kind="Function",
                name=method_name,
                file_path=file_path,
                line_start=func_def.start_point[0] + 1,
                line_end=func_def.end_point[0] + 1,
                language=language,
                parent_name=class_name,
                params=params,
            )
        )
        edges.append(
            EdgeInfo(
                kind="CONTAINS",
                source=parser._qualify(class_name, file_path, None),
                target=qualified,
                file_path=file_path,
                line=func_def.start_point[0] + 1,
            )
        )
        parser._extract_from_tree(
            func_def,
            source,
            language,
            file_path,
            nodes,
            edges,
            enclosing_class=class_name,
            enclosing_func=method_name,
            import_map=import_map,
            defined_names=defined_names,
        )

File: crates/dagayn-parser/src/r.rs
use std::collections::HashMap;

use serde_json::json;

use super::types::{ParsedEdge, ParsedNode};
use super::util::{is_test_file, line_count, node_text, strip_matching_quotes};
use super::{add_tested_by_edges, is_test_function, qualify};

pub(super) fn parse_r_with_parser(
    file_path: &str,
    source: &[u8],
    parser: Option<&mut tree_sitter::Parser>,
) -> (Vec<ParsedNode>, Vec<ParsedEdge>) {
    let line_end = line_count(source);
    let mut nodes = vec![ParsedNode {
        kind: crate::core::types::NodeKind::File,
        name: file_path.to_string(),
        file_path: file_path.to_string(),
        line_start: 1,
        line_end,
        language: "r".to_string(),
        parent_name: None,
        params: None,
        return_type: None,
        modifiers: None,
        is_test: is_test_file(file_path),
        extra: json!({}),
    }];
    let mut edges = Vec::new();
    let context = RParseContext { source, file_path };

    if let Some(parser) = parser {
        if let Some(tree) = parser.parse(source, None) {
            r_walk_children(
                tree.root_node(),
                &context,
                None,
                None,
                &mut nodes,
                &mut edges,
            );
            let mut edges = resolve_r_call_targets(&nodes, edges, file_path);
            add_tested_by_edges(&nodes, &mut edges);
            return (nodes, edges);
        }
    }

    (nodes, edges)
}

struct RParseContext<'a> {
    source: &'a [u8],
    file_path: &'a str,
}

fn r_walk_children(
    node: tree_sitter::Node<'_>,
    context: &RParseContext<'_>,
    enclosing_class: Option<&str>,
    enclosing_func: Option<&str>,
    nodes: &mut Vec<ParsedNode>,
    edges: &mut Vec<ParsedEdge>,
) {
    let mut cursor = node.walk();
    for child in node.children(&mut cursor) {
        match child.kind() {
            "binary_operator"
                if r_handle_binary_operator(child, context, enclosing_class, nodes, edges) =>
            {
                continue;
            }
            "call"
                if r_handle_call(
                    child,
                    context,
                    enclosing_class,
                    enclosing_func,
                    nodes,
                    edges,
                ) =>
            {
                continue;
            }
            _ => {}
        }
        r_walk_children(
            child,
            context,
            enclosing_class,
            enclosing_func,
            nodes,
            edges,
        );
    }
}

fn r_handle_binary_operator(
    node: tree_sitter::Node<'_>,
    context: &RParseContext<'_>,
    enclosing_class: Option<&str>,
    nodes: &mut Vec<ParsedNode>,
    edges: &mut Vec<ParsedEdge>,
) -> bool {
    let Some((left, operator, right)) = r_binary_operator_parts(node) else {
        return false;
    };
    if !matches!(operator.kind(), "<-" | "=") || left.kind() != "identifier" {
        return false;
    }
    let name = node_text(left, context.source);
    if right.kind() == "function_definition" {
        r_emit_function(right, context, &name, enclosing_class, nodes, edges);
        r_walk_children(right, context, enclosing_class, Some(&name), nodes, edges);
        return true;
    }
    if right.kind() == "call" {
        if let Some(call_name) = r_call_name(right, context.source) {
            if matches!(
                call_name.as_str(),
                "setRefClass" | "setClass" | "setGeneric"
            ) {
                r_emit_class_call(right, context, Some(&name), enclosing_class, nodes, edges);
                return true;
            }
        }
    }
    false
}

fn r_handle_call(
    node: tree_sitter::Node<'_>,
    context: &RParseContext<'_>,
    enclosing_class: Option<&str>,
    enclosing_func: Option<&str>,
    nodes: &mut Vec<ParsedNode>,
    edges: &mut Vec<ParsedEdge>,
) -> bool {
    let Some(call_name) = r_call_name(node, context.source) else {
        return false;
    };

    if matches!(call_name.as_str(), "library" | "require" | "source") {
        if let Some(target) = r_import_target(node, context.source) {
            edges.push(ParsedEdge {
                kind: crate::core::types::EdgeKind::ImportsFrom,
                source: context.file_path.to_string(),
                target,
                file_path: context.file_path.to_string(),
                line: node.start_position().row as i64 + 1,
                extra: json!({}),
            });
        }
        return true;
    }

    if matches!(
        call_name.as_str(),
        "setRefClass" | "setClass" | "setGeneric"
    ) {
        r_emit_class_call(node, context, None, enclosing_class, nodes, edges);
        return true;
    }

    r_emit_call(
        node,
        context,
        &call_name,
        enclosing_class,
        enclosing_func,
        edges,
    );
    r_walk_children(node, context, enclosing_class, enclosing_func, nodes, edges);
    true
}

fn r_emit_function(
    node: tree_sitter::Node<'_>,
    context: &RParseContext<'_>,
    name: &str,
    enclosing_class: Option<&str>,
    nodes: &mut Vec<ParsedNode>,
    edges: &mut Vec<ParsedEdge>,
) {
    let is_test = is_test_function(name, context.file_path, node, context.source);
    let qualified = qualify(context.file_path, name, enclosing_class);
    nodes.push(ParsedNode {
        kind: if is_test {
            crate::core::types::NodeKind::Test
        } else {
            crate::core::types::NodeKind::Function
        },
        name: name.to_string(),
        file_path: context.file_path.to_string(),
        line_start: node.start_position().row as i64 + 1,
        line_end: node.end_position().row as i64 + 1,
        language: "r".to_string(),
        parent_name: enclosing_class.map(str::to_string),
        params: r_direct_child_text(node, context.source, &["parameters"]),
        return_type: None,
        modifiers: None,
        is_test,
        extra: json!({}),
    });
    edges.push(ParsedEdge {
        kind: crate::core::types::EdgeKind::Contains,
        source: enclosing_class
            .map(|class| qualify(context.file_path, class, None))
            .unwrap_or_else(|| context.file_path.to_string()),
        target: qualified,
        file_path: context.file_path.to_string(),
        line: node.start_position().row as i64 + 1,
        extra: json!({}),
    });
}

fn r_emit_class_call(
    node: tree_sitter::Node<'_>,
    context: &RParseContext<'_>,
    assigned_name: Option<&str>,
    enclosing_class: Option<&str>,
    nodes: &mut Vec<ParsedNode>,
    edges: &mut Vec<ParsedEdge>,
) {
    let Some(class_name) = r_first_string_arg(node, context.source).or_else(|| {
        assigned_name
            .filter(|name| !name.is_empty())
            .map(str::to_string)
    }) else {
        return;
    };
    let qualified = qualify(context.file_path, &class_name, enclosing_class);
    nodes.push(ParsedNode {
        kind: crate::core::types::NodeKind::Class,
        name: class_name.clone(),
        file_path: context.file_path.to_string(),
        line_start: node.start_position().row as i64 + 1,
        line_end: node.end_position().row as i64 + 1,
        language: "r".to_string(),
        parent_name: enclosing_class.map(str::to_string),
        params: None,
        return_type: None,
        modifiers: None,
        is_test: false,
        extra: json!({}),
    });
    edges.push(ParsedEdge {
        kind: crate::core::types::EdgeKind::Contains,
        source: context.file_path.to_string(),
        target: qualified,
        file_path: context.file_path.to_string(),
        line: node.start_position().row as i64 + 1,
        extra: json!({}),
    });
    if let Some(methods) = r_find_named_arg(node, context.source, "methods") {
        r_extract_methods(methods, context, &class_name, nodes, edges);
    }
}

fn r_extract_methods(
    list_call: tree_sitter::Node<'_>,
    context: &RParseContext<'_>,
    class_name: &str,
    nodes: &mut Vec<ParsedNode>,
    edges: &mut Vec<ParsedEdge>,
) {
    for (method_name, value) in r_iter_args(list_call, context.source) {
        let Some(method_name) = method_name else {
            continue;
        };
        if value.kind() != "function_definition" {
            continue;
        }
        r_emit_function(value, context, &method_name, Some(class_name), nodes, edges);
        r_walk_children(
            value,
            context,
            Some(class_name),
            Some(&method_name),
            nodes,
            edges,
        );
    }
}

fn r_emit_call(
    node: tree_sitter::Node<'_>,
    context: &RParseContext<'_>,
    call_name: &str,
    enclosing_class: Option<&str>,
    enclosing_func: Option<&str>,
    edges: &mut Vec<ParsedEdge>,
) {
    let caller = enclosing_func
        .map(|func| qualify(context.file_path, func, enclosing_class))
        .unwrap_or_else(|| context.file_path.to_string());
    edges.push(ParsedEdge {
        kind: crate::core::types::EdgeKind::Calls,
        source: caller.clone(),
        target: call_name.to_string(),
        file_path: context.file_path.to_string(),
        line: node.start_position().row as i64 + 1,
        extra: json!({}),
    });
    if let Some(edge) = r_bridge_edge(node, context, &caller, call_name) {
        edges.push(edge);
    }
}

fn r_bridge_edge(
    node: tree_sitter::Node<'_>,
    context: &RParseContext<'_>,
    caller: &str,
    signature: &str,
) -> Option<ParsedEdge> {
    let (relationship_role, bridge_kind) = match signature {
        "system" | "system2" => ("invokes_binary", "subprocess"),
        ".Call" | ".External" => ("loads_native_module", "ffi"),
        "dyn.load" | "library.dynam" => ("loads_shared_library", "ffi"),
        "readLines" | "read.csv" | "read.table" => ("reads_file", "file_io"),
        "writeLines" | "write.csv" => ("writes_file", "file_io"),
        _ => return None,
    };
    let line = node.start_position().row as i64 + 1;
    let (target, confidence, confidence_tier) = match r_first_string_arg(node, context.source) {
        Some(target) => (target, 0.8, "HIGH"),
        None => (
            format!("<dynamic:{signature}@{}:{line}>", context.file_path),
            0.2,
            "LOW",
        ),
    };
    Some(ParsedEdge {
        kind: crate::core::types::EdgeKind::CrossArtifact,
        source: caller.to_string(),
        target,
        file_path: context.file_path.to_string(),
        line,
        extra: json!({
            "relationship_role": relationship_role,
            "bridge_kind": bridge_kind,
            "evidence_kind": "syntax",
            "evidence_source": signature,
            "source_language": "r",
            "target_language": "unknown",
            "confidence": confidence,
            "confidence_tier": confidence_tier,
        }),
    })
}

fn r_binary_operator_parts<'a>(
    node: tree_sitter::Node<'a>,
) -> Option<(
    tree_sitter::Node<'a>,
    tree_sitter::Node<'a>,
    tree_sitter::Node<'a>,
)> {
    let mut cursor = node.walk();
    let children = node.children(&mut cursor).collect::<Vec<_>>();
    if children.len() < 3 {
        return None;
    }
    Some((children[0], children[1], children[2]))
}

fn r_call_name(node: tree_sitter::Node<'_>, source: &[u8]) -> Option<String> {
    let mut cursor = node.walk();
    for child in node.children(&mut cursor) {
        if matches!(child.kind(), "identifier" | "namespace_operator") {
            return Some(node_text(child, source));
        }
    }
    None
}

fn r_import_target(node: tree_sitter::Node<'_>, source: &[u8]) -> Option<String> {
    let (_, value) = r_iter_args(node, source).into_iter().next()?;
    match value.kind() {
        "identifier" => Some(node_text(value, source)),
        "string" => r_string_text(value, source),
        _ => None,
    }
}

fn r_first_string_arg(node: tree_sitter::Node<'_>, source: &[u8]) -> Option<String> {
    let (_, value) = r_iter_args(node, source).into_iter().next()?;
    if value.kind() == "string" {
        r_string_text(value, source)
    } else {
        None
    }
}

fn r_find_named_arg<'a>(
    node: tree_sitter::Node<'a>,
    source: &[u8],
    arg_name: &str,
) -> Option<tree_sitter::Node<'a>> {
    r_iter_args(node, source)
        .into_iter()
        .find_map(|(name, value)| (name.as_deref() == Some(arg_name)).then_some(value))
}

fn r_iter_args<'a>(
    call_node: tree_sitter::Node<'a>,
    source: &[u8],
) -> Vec<(Option<String>, tree_sitter::Node<'a>)> {
    let Some(arguments) = r_direct_child(call_node, &["arguments"]) else {
        return Vec::new();
    };
    let mut out = Vec::new();
    let mut cursor = arguments.walk();
    for argument in arguments.children(&mut cursor) {
        if argument.kind() != "argument" {
            continue;
        }
        let mut name = None;
        let mut value = None;
        let mut seen_equals = false;
        let mut arg_cursor = argument.walk();
        for child in argument.children(&mut arg_cursor) {
            if child.kind() == "=" {
                seen_equals = true;
                continue;
            }
            if !child.is_named() {
                continue;
            }
            if seen_equals {
                value = Some(child);
                break;
            }
            if name.is_none() && child.kind() == "identifier" {
                name = Some(node_text(child, source));
                continue;
            }
            if value.is_none() {
                value = Some(child);
                break;
            }
        }
        if let Some(value) = value.or_else(|| r_first_named_child(argument)) {
            out.push((seen_equals.then_some(name).flatten(), value));
        }
    }
    out
}

fn r_string_text(node: tree_sitter::Node<'_>, source: &[u8]) -> Option<String> {
    r_first_descendant_text(node, source, &["string_content"])
        .or_else(|| Some(strip_matching_quotes(node_text(node, source).trim()).to_string()))
        .filter(|value| !value.is_empty())
}

fn r_direct_child<'a>(
    node: tree_sitter::Node<'a>,
    kinds: &[&str],
) -> Option<tree_sitter::Node<'a>> {
    let mut cursor = node.walk();
    let found = node
        .children(&mut cursor)
        .find(|child| kinds.contains(&child.kind()));
    found
}

fn r_direct_child_text(
    node: tree_sitter::Node<'_>,
    source: &[u8],
    kinds: &[&str],
) -> Option<String> {
    r_direct_child(node, kinds).map(|child| node_text(child, source))
}

fn r_first_named_child<'a>(node: tree_sitter::Node<'a>) -> Option<tree_sitter::Node<'a>> {
    let mut cursor = node.walk();
    let found = node.children(&mut cursor).find(|child| child.is_named());
    found
}

fn r_first_descendant_text(
    node: tree_sitter::Node<'_>,
    source: &[u8],
    kinds: &[&str],
) -> Option<String> {
    let mut cursor = node.walk();
    for child in node.children(&mut cursor) {
        if kinds.contains(&child.kind()) {
            return Some(node_text(child, source));
        }
        if let Some(found) = r_first_descendant_text(child, source, kinds) {
            return Some(found);
        }
    }
    None
}

fn resolve_r_call_targets(
    nodes: &[ParsedNode],
    edges: Vec<ParsedEdge>,
    file_path: &str,
) -> Vec<ParsedEdge> {
    let symbols = nodes
        .iter()
        .filter(|node| matches!(node.kind.as_str(), "Function" | "Test"))
        .fold(HashMap::<String, String>::new(), |mut symbols, node| {
            symbols
                .entry(node.name.clone())
                .or_insert_with(|| qualify(file_path, &node.name, node.parent_name.as_deref()));
            symbols
        });
    edges
        .into_iter()
        .map(|mut edge| {
            if edge.kind == "CALLS" && !edge.target.contains("::") {
                if let Some(target) = symbols.get(&edge.target) {
                    edge.target = target.clone();
                }
            }
            edge
        })
        .collect()
}
